### PR TITLE
Updates for latest revision

### PR DESCRIPTION
**Pull request by RAMitchell**

### CODE REVIEW

File: include/shuffle/BijectiveFunctionScanShuffle.h [hunk 1]
```diff
@@ -3,149 +3,209 @@
...
-    }
+  char *allocate(std::ptrdiff_t num_bytes) {
+    memory.resize(num_bytes);
+    return memory.data().get();
+  }
 
-    void deallocate( char* ptr, size_t )
-    {
-    }
+  void deallocate(char *ptr, size_t) {}
```
Comment (djns99) [inline include/shuffle/BijectiveFunctionScanShuffle.h]: Can you run clang format with the .clang-format file in the root of the project. This should make it easier to see what you have changed.

File: test/MMD.cu [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            for( auto alpha : thresholds )
+            {
+                for( auto round = begin_round; round < end_round; round++ )
+                {
+                    std::cout << "$\\alpha_H=" << alpha << "$,";
+                    printf( "%f,%d,%d,%d\n", HoeffdingAcceptanceThreshold( alpha, num_samples ),
+                            int( num_samples ), int( n ), int( round ) );
+                    std::cout << "$\\alpha_N=" << alpha << "$,";
+                    printf( "%f,%d,%d,%d\n", NormalAcceptanceThreshold( alpha, num_samples, n, lambda ),
+                            int( num_samples ), int( n ), int( round ) );
```
Comment (djns99) [inline test/MMD.cu:303]: Can you add the raw data files somewhere so we have them for posterity

**Review by djns99 — COMMENTED** [review]
Looks good for the most part. Just a couple of minor requests

**Comment by djns99** [comment]
Looks good for the most part. I would ask if maybe we could keep most of the supplementary files etc. in a separate dev branch so we have them for reference. This would be particularly important if we need to wipe the commit history to get rid of the large files. I would also want to keep the tests around on a development branch in case we/some future work wants to come back to this in the future with new bijective functions.

Just to confirm, do you have all the results that went into the paper available? It looks like the MMD ones dont create a file?
I think there is some value to have some raw data files available (as opposed to just graphs) so that if people want to reproduce or compare results they have something precise to refer to. My personal opinion is where possible we should be as transparent and helpful as possible (even to just our future selves).

Perhaps it may be worth merging these changes into a separate release branch that we tag as the paper version. My concern is stripping everything back like this gives us very limited flexibility if we ever want to revisit the project. In particular I think a couple of the implementations such as the Czumaj network one have room to be improved substantially in future if we ever want to come back to it.

**Comment by RAMitchell** [comment]
Definitely agree with keeping the implementations we didn't necessarily use somewhere in version control. The 'front facing' version should correspond only to the material in the paper however. I'll leave it up to you how you want to arrange that.

I don't have time to sort out MMD experiments to produce a file at the moment, but yes I probably should have done it that way.

**Comment by djns99** [comment]
Ok I will mess around with the branches and sort out a dev branch with all the in progress stuff. Are the MMD tests deterministic, if so then it probably doesnt matter too much that there isn't a file

**Comment by RAMitchell** [comment]
Yes they are deterministic.

**Pull request merged**